### PR TITLE
cmake: add support for out of src tree builds

### PR DESCRIPTION
**Pull request by russkel**
I was attempting to build Core from a read-only volume in Docker and it failed because part of the build process generates files directly into the source tree (not really recommended practice). This PR fixes that by generating the files into the build tree instead (if you build into the source tree, then source and build tree are the same directory). I have tested both in source and out of source builds and both work as expected.

I also took the opportunity to fix MOOSVersion.h by merging with GitVersion.h and populating it from the CMakeLists package var, as the file was badly out of date.

Git SHA hashes are also typically truncated to 7 characters (that's what git uses).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -239,6 +239,9 @@ install(
 # Save a version file in the project's install directory
 install(FILES ${VERSION_FILE} DESTINATION "${PROJECT_CONFIG_PATH}")
 
+# Install the MOOSVersion header
+install(FILES ${CMAKE_BINARY_DIR}/Core/libMOOS/include/MOOS/libMOOS/MOOSVersion.h DESTINATION include/MOOS/libMOOS)
```
Comment (alastair-harrison-oxb) [inline CMakeLists.txt:243]: The first path probably ought to be double-quoted, in case of spaces in `CMAKE_BINARY_DIR`.

File: Core/libMOOS/CMakeLists.txt [hunk 2]
```diff
@@ -103,6 +89,7 @@ set(INCLUDE_ROOTS
     ${CMAKE_CURRENT_SOURCE_DIR}/Thirdparty/PocoBits/include
     ${CMAKE_CURRENT_SOURCE_DIR}/Thirdparty/getpot/include
     ${CMAKE_CURRENT_SOURCE_DIR}/Thirdparty/AppCasting/include
+    ${CMAKE_CURRENT_BINARY_DIR}/include
```
Comment (alastair-harrison-oxb) [inline Core/libMOOS/CMakeLists.txt:92]: I think this path should be double-quoted too. And those above it.

File: Core/libMOOS/CMakeLists.txt [hunk 3]
```diff
@@ -199,6 +193,13 @@ install(
     FILES_MATCHING PATTERN "*.h" PATTERN "*.hxx" PATTERN "*.hpp"
 )
 
+# Create and install ClientDefines.h header for ASYNCHRONOUS_CLIENT_DEFINE variable
+configure_file(${CMAKE_MODULE_PATH}/ClientDefines.h.in
```
Comment (alastair-harrison-oxb) [inline Core/libMOOS/CMakeLists.txt:197]: These paths should be quoted too. I know the originals weren't, and I know that it was probably me who wrote the originals, but you've got to start somewhere, right :-)

We can make a separate PR to fix the the rest of them, but it seems right to fix these one here while you're touching them.

File: Core/libMOOS/CMakeLists.txt [hunk 4]
```diff
@@ -199,6 +193,13 @@ install(
...
+    ${CMAKE_CURRENT_BINARY_DIR}/include/MOOS/libMOOS/App/ClientDefines.h
+    @ONLY
+)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/include/MOOS/libMOOS/App/ClientDefines.h DESTINATION include/MOOS/libMOOS/App)
```
Comment (alastair-harrison-oxb) [inline Core/libMOOS/CMakeLists.txt:201]: Quote paths here too.

**Comment by russkel** [comment]
Sorry about not splitting them up into separate commits. I kind of did the work all in one go to get it working for my uses.

I have made the modifications as you have requested, but is this actually an issue with CMake? I'm not sure CMake will do literal expansion of variables in this manner.

Thanks for the review, @aharrison24

**Comment by aharrison24** [comment]
I've done some experiments and you are completely correct - the quoting of paths in function arguments is not necessary (even if they contain spaces). The only place it does appear to be necessary is for functions that take shell commands as arguments.

I also checked in Craig Scott's "Professional CMake" book and couldn't find any mention of path quoting. If he doesn't talk about it, then it probably isn't necessary!

Thanks for teaching me a new thing, and sorry for causing you to do extra work.

**Comment by russkel** [comment]
> Thanks for teaching me a new thing, and sorry for causing you to do extra work.

No problem, was two seconds. Explicit is better than implicit, as they say.

**Comment by pmnewman** [comment]
thank you
